### PR TITLE
Authentication

### LINKED ISSUE #11: Authentication
#### As an application hoster I want to add user authentication to my application.

- [ ] Users can sign up.
- [ ] Users can log in after they've signed up.
- [ ] Users can log out after logging in.
- [ ] Users can choose to be remembered after logging in.<sup>†</sup>
- [ ] Multiple browsers can concurrently hold permanent cookies for the same
  user.
- [ ] Logged in users should not be able to register, nor log in again.

<sup>†</sup>: "Remembering" means that users will remain logged into the site
even after completely closing their web browser.

__Note__: Do not use Devise for this task.

__Discussion topics__:
- [ ] If we wanted to add more advanced features such as email activation and
  "forgot my password", should we continue to "roll our own" authentication?

__Reference material__:
- https://www.railstutorial.org/book/log_in_log_out

__Further reading__:
- https://github.com/plataformatec/devise

__Dependencies__:
- Delete Image

### PR DESCRIPTION
**Pull request by tlconnor**
Closes #11

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -21,6 +21,44 @@
     t.datetime "updated_at", null: false
   end
 
+  create_table "oauth_access_grants", force: :cascade do |t|
```
Comment (bboe) [inline db/schema.rb]: Where did all this come from?

File: test/models/user_test.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  def test_invalid_email
+    user = User.new(email: 'invalid')
+    assert_predicate user, :invalid?
+    assert_equal ['is invalid'], user.errors[:email]
+  end
+
+  def test_unique_email
+    User.create!(email: 'nobody@appfolio.com', password: 'password1', first_name: 'First', last_name: 'Last')
+
+    user = User.new(email: 'nobody@appfolio.com')
```
Comment (bboe) [inline test/models/user_test.rb]: What about `Nobody@appfolio.com`?

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  validates :first_name, presence: true
+  validates :last_name, presence: true
+
+  validates :email, presence: true, uniqueness: {
+    message: 'has already been taken'
+  }, format: { with: /.+@.+\..+/i, allow_blank: true }
```
Comment (bboe) [inline app/models/user.rb]: This is both very permissive in some areas and restrictive in others (doesn't appear to permit emails for subdomains). Can we leverage the regex (or whatever you're using) that validates emails for share user here as well?

File: Gemfile [hunk 4]
```diff
@@ -53,6 +53,7 @@ end
 group :test do
   gem 'capybara'
   gem 'selenium-webdriver'
+  gem 'mocha'
```
Comment (bboe) [inline Gemfile]: [minor] Can we try to keep lists of similar items like this sorted?

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+class UsersController < ApplicationController
+  before_action :doorkeeper_authorize!, except: [:create]
+
+  def create
+    @user = User.new(params.permit(:first_name, :last_name, :email, :password, :password_confirmation))
```
Comment (bboe) [inline app/controllers/users_controller.rb]: Do we care if the user completely omits the password_confirmation field? Based on the model tests, that field doesn't seem required.

Comment (bboe) [inline app/controllers/users_controller.rb]: Also, consider `User.create!`.

File: test/controllers/users_controller_test.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        password_confirmation: 'password1',
+        first_name: 'First',
+        last_name: 'Last'
+      }
+    end
+
+    assert_response :success
+    user = User.last
+    body = JSON.parse(response.body)
+    assert_equal 201, response.status
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: This seems redundant to `assert_response`. Can you instead do `assert_response :created`?

File: test/controllers/users_controller_test.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
+require 'test_helper'
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
+  test 'should post create' do
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: What if you're already logged in?

File: test/controllers/users_controller_test.rb [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    token = mock(acceptable?: true)
+    user = User.new(
+      email: 'nobody@appfolio.com', password: 'password1', first_name: 'First', last_name: 'Last'
+    )
+    UsersController.any_instance.stubs(:doorkeeper_token).returns(token)
+    UsersController.any_instance.stubs(:current_user).returns(user)
+
+    get me_users_url, as: :json
+    body = JSON.parse(response.body)
+    assert_equal 200, response.status
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: prefer using `assert_response` with the named statuses, rather than their number.

File: test/controllers/users_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    assert_equal 200, body['status']
+    assert_equal user.id, body['user']['id']
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Can we dry up these assertions? They are similar to those in create.

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+    @user.save!
+    render status: 201
+  end
+
+  def me
```
Comment (bboe) [inline app/controllers/users_controller.rb]: What's the purpose of `me` in your application?

File: test/components/authentication.spec.js [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+  describe('#getUser', () => {
+    it('returns the locally stored user', () => (
+      authentication
+        .authenticate('nobody@appfolio.com', 'password1')
+        .then(() => {
+          expect(authentication.getUser()).to.deep.equal(user);
+        })
+    ));
+  });
+});
```
Comment (bboe) [inline test/components/authentication.spec.js]: Great tests.

File: app/models/user.rb [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+
+  has_secure_password
```
Comment (bboe) [inline app/models/user.rb]: Should we have a minimum password length?

**Comment by bboe** [comment]
I think that's all my comments for the first round.

File: db/schema.rb (hunk 1 above)
Comment (tlconnor) [inline db/schema.rb]: Ah, looks like I messed up the splitting of the commits.  The OAuth related DB changes should have been part of https://github.com/AppFolioOnboarding/image-sharer-tlconnor/pull/31/commits/25cdfda92f7ac40b65896c2e8b5ad6b440d329f3

I won't go change it right now because it will mess up all of your comments on the PR but I could rebase at the end of the review to fix it up.

File: test/models/user_test.rb (hunk 2 above)
Comment (tlconnor) [inline test/models/user_test.rb]: Good point, fixed (and added `case_sensitive: false` on the validation)

File: app/models/user.rb (hunk 3 above)
Comment (tlconnor) [inline app/models/user.rb]: You are wrong about not permitting emails from subdomains, the `.+` allows anything including multiple subdomains.  All this regex is checking is that the email includes a `username`, then `@` then `anything` then `.` then `anything`.

This is the exact same regex as per the `share image` story so you are right that they could share the same regex.  Once one of the PRs has been merged the other PR could be updated to refactor the regex to be common between both.

File: app/controllers/users_controller.rb (hunk 5 above)
Comment (tlconnor) [inline app/controllers/users_controller.rb]: I've just gone with the default behavior of `has_secure_password` as defined here http://api.rubyonrails.org/classes/ActiveModel/SecurePassword/ClassMethods.html, so the `password_confirmation` is only checked if provided.  I'm ok with this.

I've switched to `User.create!` as suggested.

File: Gemfile (hunk 4 above)
Comment (tlconnor) [inline Gemfile]: fixed

File: test/controllers/users_controller_test.rb (hunk 6 above)
Comment (tlconnor) [inline test/controllers/users_controller_test.rb]: I didn't know about `assert_response :created`.  I'll use that in the future (and fixed here)

File: test/controllers/users_controller_test.rb (hunk 7 above)
Comment (tlconnor) [inline test/controllers/users_controller_test.rb]: I don't see that as being relevant.  This is an API to create a user, there is nothing saying that the user happens to be me.

File: test/controllers/users_controller_test.rb (hunk 8 above)
Comment (tlconnor) [inline test/controllers/users_controller_test.rb]: Fixed

File: test/controllers/users_controller_test.rb (hunk 9 above)
Comment (tlconnor) [inline test/controllers/users_controller_test.rb]: Fixed

File: app/controllers/users_controller.rb (hunk 10 above)
Comment (tlconnor) [inline app/controllers/users_controller.rb]: After logging in the client has an auth token but doesn't have any information about the currently logged in user.  The `/users/me` API endpoint returns the details of the current user so we can show their name, email, etc in the UI.

File: app/models/user.rb (hunk 12 above)
Comment (tlconnor) [inline app/models/user.rb]: In a real world app we would want to test for length, inclusion of character groups, avoiding previous passwords,  etc, etc whereas here I went for the minimum effort to meet the task requirements.

For the sake of the exercise I've added minimum length, though in each task in ropes I've wondered how far I really need to take them.

File: test/controllers/users_controller_test.rb (hunk 6 above)
Comment (bboe) [inline test/controllers/users_controller_test.rb]: `assert_response` can be used with all the status codes. The numerical values can also be used, but nearly all of them are mapped to some name. I am actually surprised that `:success` works for all of the 200-level responses. I suppose `:okay` is the one specifically for `200`.

File: test/controllers/users_controller_test.rb (hunk 7 above)
Comment (bboe) [inline test/controllers/users_controller_test.rb]: That's a good point, but could it be possible that through using multiple browser tabs the behavior of your app could enter an unsupported state transition by allowing this interaction?

As an example (doesn't directly apply), say I have two tabs open to the login page. On tab 1 I login as my primary account, and then sometime later I forget about tab 1, and login on tab 2 as a different user. Then sometime later I go back to tab 1 and think I'm logged in as my primary account. When I click on a link to perform an action it suddenly appears that I am logged in as my secondary user. This state change transparently from user1 to user2 enables this behavior and is not something we want to support.

You're right that user creation in itself isn't a problem, unless your app changes the "session's" state along with the creation of a user.

File: app/controllers/users_controller.rb (hunk 10 above)
Comment (bboe) [inline app/controllers/users_controller.rb]: Cool.

File: app/models/user.rb (hunk 12 above)
Comment (bboe) [inline app/models/user.rb]: Yes, you're right, we haven't really specified. Part of that is to allow some flexibility and creativity, and part of it is to see what things everyone feels are reasonable minimums for a feature.

File: app/controllers/users_controller.rb [hunk 13]
```diff
@@ -0,0 +1,12 @@
+class UsersController < ApplicationController
+  before_action :doorkeeper_authorize!, except: [:create]
+
+  def create
+    @user = User.create!(params.permit(:first_name, :last_name, :email, :password, :password_confirmation))
+    render status: 201
```
Comment (bboe) [inline app/controllers/users_controller.rb]: Can we used the named value rather than `201`?

File: app/views/users/create.json.jbuilder [hunk 14]
```diff
@@ -0,0 +1,2 @@
+json.status 201
```
Comment (bboe) [inline app/views/users/create.json.jbuilder]: Does jbuilder permit named status codes? If so, can we use them instead of the numbers?

File: test/controllers/users_controller_test.rb [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+      email: 'nobody@appfolio.com',
+      password: 'password1',
+      first_name: 'First',
+      last_name: 'Last'
+    )
+    UsersController.any_instance.stubs(:doorkeeper_token).returns(token)
+    UsersController.any_instance.stubs(:current_user).returns(user)
+
+    get me_users_url, as: :json
+    assert_response :success
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Can you use the more precise `:ok` here if it should be a 200?

File: test/models/user_test.rb [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+  end
+
+  def test_blank_password
+    user = User.new
+    assert_predicate user, :invalid?
+    assert_equal ["can't be blank"], user.errors[:password]
+  end
+
+  def test_short_password
+    user = User.new(password: 'a')
```
Comment (bboe) [inline test/models/user_test.rb]: When testing min/maxes I think it's a good idea to have the test check the edge case. Here a 7-digit password would be perfect. Similarly the success case should use an 8 digit one for the success edge-case.

**Comment by bboe** [comment]
From the QA perspective, I was testing multiple tabs, and it seems that my session is only valid for a single tab. Opening a new tab does not keep my session. Is there a way to handle that?

**Comment by bboe** [comment]
Also, I'm not sure if it has anything to do with this PR, but in the console I see:

image

File: app/views/users/create.json.jbuilder (hunk 14 above)
Comment (rmacklin) [inline app/views/users/create.json.jbuilder]: I'm confused why this is here at all. Seems redundant to put the status code in the response body.

**Comment by rmacklin** [comment]
I noticed that in the signup form I couldn't press enter to submit the form. I didn't check the other forms, but that's a UX antipattern.

File: app/views/users/create.json.jbuilder (hunk 14 above)
Comment (bboe) [inline app/views/users/create.json.jbuilder]: Not that Richard commented, I remember that you're manually adding the response status to the json body similar to how many other json APIs do.

Comment (rmacklin) [inline app/views/users/create.json.jbuilder]: With it done manually like this, it's very error prone that a developer could change one and forget to change the other. It would be best to DRY it so it's impossible for them to get out of sync

**Comment by tlconnor** [comment]
@bboe @rmacklin I've rebased this branch against the latest master and resolved all issues mentioned above.

**Comment by bboe** [comment]
image

Once I get that error message, it doesn't automatically go away as I browse through other "pages" as I would expect it to.

**Comment by bboe** [comment]
The app doesn't prevent me from logging in when I'm already logged in.

**Comment by bboe** [comment]
image

I'm not sure if it's an easy problem to fix. When I refresh the page manually, I see the "login / register" first and then see it change to what is shown above. Perhaps an overlay can be used while the page is "loading" to avoid seeing that odd transition?

**Comment by bboe** [comment]
All other functionality looks good!

**Comment by tlconnor** [comment]
@bboe I've submitted https://github.com/AppFolioOnboarding/image-sharer-tlconnor/pull/42 as a fix for the message bar not dismissing

**Comment by tlconnor** [comment]
@bboe all reported issues are now fixed:

- https://github.com/AppFolioOnboarding/image-sharer-tlconnor/pull/42 fixes the message bar dismissal
- login and signup screens will now redirect if the user is already logged in
- a loading spinner has been added for the initial state when the user details are being loaded

**Comment by bboe** [comment]
Very much digging the spinner. Surprisingly, I don't actually see it unless I'm looking for it, but for whatever reason I always saw the transition before.

Also, I like that I cannot login as another user when I'm already logged in. That satisfies avoiding the invalid state change. It would be nice if there was a message to the effect of "please logout before trying to login", however, since it should be a rare-case I'm okay if you don't add that.

👍  Great work.
